### PR TITLE
fix(v3/linux): confine remaining WebKit request ops to the GTK main thread (#5631)

### PR DESCRIPTION
**Pull request by taliesin-ai**
## Summary

Follow-up to #5640 for the long-running Linux WebKitGTK crashes in #5631.

#5640 marshalled the response completion (`webkit_uri_scheme_request_finish_*`) onto the GTK main thread via `invokeOnMainSync`, but **deliberately left the request metadata reads and the request-body input stream on the asset-server worker goroutine**. `v3.0.0-alpha2.105` ships #5640, and the reporter confirms it *still* aborts in `g_object_unref` during `g_main_context_iteration` (~14 aborts / 17h). After confining the remaining operations, they observed ~22h continuously alive.

WebKit2GTK is not thread-safe — and neither GTK3 nor GTK4 is. **Every** call that touches a WebKit/libsoup object owned by the GTK main loop must run on the thread driving `g_application_run`. The asset server completes requests from a pool of worker goroutines, so the calls #5640 left off-thread are still racing the main loop. The crash lands in `g_object_unref` because dropping the last ref triggers *finalization* (which touches main-loop state) — the atomic refcount itself is fine; the finalizer is what races. This is identical in webkit2gtk-4.1 (GTK3) and webkitgtk-6.0 (GTK4), so the fix applies to both builds even though the report is from a GTK3 deployment.

## What this changes

**1. Confine the remaining WebKit operations** to the GTK main thread via the existing `invokeOnMainSync` (`mainthread_linux.go`, added in #5640):

- `request.URL()` — `webkit_uri_scheme_request_get_uri`
- HTTP method read — `webkit_uri_scheme_request_get_http_method`
- HTTP header read + libsoup iteration — `webkit_uri_scheme_request_get_http_headers`
- request body lookup — `webkit_uri_scheme_request_get_http_body`
- request body stream `Read` (`g_input_stream_read_all`) and `Close` (`g_input_stream_close` + `g_object_unref`)
- the **response `GUnixInputStream`**: now created *and* unref'd inside the same `invokeOnMainSync` hop as `webkit_uri_scheme_response_new` / `finish_with_response`. Previously it was built and unref'd on the worker while WebKit took its ref on the main thread — a cross-thread refcount split on the stream. `webkit_uri_scheme_request_finish` now takes the pipe FD instead of a pre-built stream.

The bulk response data path (writing the body to the pipe FD) intentionally stays on the worker — it touches no WebKit objects. The request unref on `Close()` keeps the existing async main-context path from #5566.

**2. Shutdown guard** (addresses jalfdev's review question 5). `invokeOnMainSync` blocks the worker on a `GCond` until the main loop services its scheduled source. If the loop stops while a worker is mid-request, the source is never run and the worker would block forever. `appRun` now calls `webview.DisableMainThreadDispatch()` the instant `g_application_run` returns; after that, `invokeOnMainSync` runs the callback **inline** on the worker instead of scheduling it. Inline is safe precisely because the loop is no longer iterating — the cross-thread race confinement guards against is gone — and in-flight asset requests drain during shutdown instead of wedging.

Applied to both the **GTK3** (`webkit2gtk-4.1`) and default **GTK4** (`webkitgtk-6.0`) cgo builds. The purego build was removed in #5661, so it no longer applies.

## Verification (Ubuntu 26.04, WebKitGTK 2.52.3 / GTK 4.22.2 & GTK 3.24, go1.26.2)

- Compiles clean and `go vet` clean on both the default (GTK4) and `-tags gtk3` builds.
- **No deadlock under load**, the real risk of adding several synchronous main-thread round-trips per request. A headless (`xvfb`) app served **4000 GETs + 4000 POSTs (6 MB of request bodies fully read via the confined body-stream path)** and completed cleanly on **both** builds. GTK3 also rendered correctly (CSS + JS served via the asset path; JS ran to completion).
- **Clean shutdown** with the guard: on both builds the app completed the load, called `Quit()`, `g_application_run` returned, and the process **exited on its own** (GTK3 fast; GTK4 ~54s under software rendering) — no shutdown wedge.

**Not verified:** the underlying race is a multi-hour soak (unreproducible in CI). This PR implements @jalfdev's analysis and reproduces the operations from their ~22h-validated patch outline, but I have **not** independently soak-tested it.

## Credit

Diagnosis and the validated patch outline are from **@jalfdev** in #5631. 🙏

Refs #5631, #5640, #5566, #5661

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved Linux webview handling to ensure request metadata, request bodies, and response finishing are processed on the required GTK/UI main loop thread.
  * Updated response completion to pass the pipe file descriptor directly for more reliable stream handling.
  * Prevented shutdown hangs by disabling main-loop dispatch after the main loop exits, avoiding blocked in-flight requests.
  * Added safeguards for zero-length request body reads to avoid edge-case panics.
* **Tests**
  * Added coverage for main-loop dispatch behavior during shutdown and for zero-length request body reads.

### CODE REVIEW

**Comment by taliesin-ai** [comment]
One tradeoff worth calling out, since it overrides a deliberate decision in #5640:

#5640 intentionally left the request-body input stream on the worker, reasoning it was "transfer-full, owned solely by us" and therefore safe off-thread. This PR confines it anyway, because alpha2.105 ships #5640 and still aborts. The cost: `g_input_stream_read_all` now runs on the GTK main thread, so a slow or very large upload body would **block the UI thread** (a transient freeze, not a crash). For a local asset server, request bodies are small and immediate (the 6 MB load test drained instantly), so this is an acceptable trade.

Note also that for the GET-heavy "frequent UI updates / media rendering" workload in #5631, the metadata reads (URL/method/headers) — which run on *every* request — are the most likely crash culprit; the body-stream confinement is more belt-and-suspenders. Happy to narrow the body path back to the worker if the main-thread-blocking concern outweighs it, but confining everything matches @jalfdev's validated patch.

File: v3/internal/assetserver/webview/mainthread_linux.go [hunk 1]
```diff
@@ -45,6 +56,17 @@ static gboolean webview_main_sync_trampoline(gpointer data) {
...
+	// scheduled source would never run and the worker would block here forever.
+	// The loop is also no longer iterating, so the cross-thread race that makes
+	// main-thread confinement necessary is gone — running the callback inline on
+	// the worker is safe at that point and lets in-flight asset requests drain
+	// during shutdown instead of wedging. See #5631 (review question 5).
+	if (!g_atomic_int_get(&webview_main_dispatch_enabled)) {
+		webviewMainThreadCallback(id);
+		return;
+	}
+
```
[the remark this answers is not in the document]
Comment (taliesin-ai) [reply v3/internal/assetserver/webview/mainthread_linux.go]: Fixed in 8d2cd1bc8. The atomic flag is replaced by a dedicated `GMutex` (`webview_dispatch_mu`) that serializes the enabled-check **and** the `g_main_context_invoke` against the flag clear in `webview_disable_main_dispatch`. A worker now holds the mutex across both, so it either schedules onto a live loop or sees the loop already stopped and runs the callback inline — it can never read `enabled` and then have the flag flip before it schedules.

The trampoline only touches the per-call mutex/cond, never `webview_dispatch_mu`, so when `g_main_context_invoke` runs it inline for a main-thread caller there is no self-deadlock.

On the deliberately *not* fixed residual: a worker can still read `enabled == TRUE`, schedule, and park if the loop happens to stop in the tiny window after `g_application_run` returns but before `DisableMainThreadDispatch` runs. This is harmless and intentional — the asset-server worker goroutines are never joined (`for req := range workerC` in `ServeWebViewRequest`), so a parked worker is reaped at process exit, and its stack-allocated `webviewMainSyncCall` stays valid the whole time (no use-after-free). Draining queued sources with `g_main_context_iteration` to close that window would dispatch *arbitrary* ready GTK/WebKit teardown sources off the main thread — re-introducing exactly the finalization race this PR exists to fix — so the trade isn't worth it.

The inline-during-shutdown path runs WebKit ops off the GTK thread, relying on "loop stopped → no concurrent main-thread access." Flagging that assumption explicitly for maintainer review.

Added a cgo test (`TestMainThreadDispatch`, commit 1b1f321a4) that drives this against a real GLib main loop: callbacks run on the loop thread while enabled, racing disable against in-flight workers never deadlocks, and post-disable calls run inline. Passes under `-race` on both gtk3 and gtk4.

**Comment by taliesin-ai** [comment]
Thanks @coderabbitai — both findings are addressed.

**🟠 Major — shutdown race on the dispatch flag** (8d2cd1bc8): the `volatile gint` atomic flag is replaced by a `GMutex` that serializes the enabled-check together with `g_main_context_invoke`, so the check and the scheduling it gates are atomic with respect to `webview_disable_main_dispatch`. Details and the rationale for the remaining (harmless, intentional) park are in the [inline reply](https://github.com/wailsapp/wails/pull/5668#discussion_r3487953086).

**🟡 Minor — zero-length `Read` panic** (8d2cd1bc8): added `if len(p) == 0 { return 0, nil }` ahead of `&p[0]` in both `webkit_linux.go` (gtk4) and `webkit_linux_gtk3.go`, matching the `io.Reader` contract.

**Test coverage** (1b1f321a4): new `TestMainThreadDispatch` drives `invokeOnMainSync`/`DisableMainThreadDispatch` against a real GLib main loop — confirms loop-thread confinement while enabled, no deadlock when disable races in-flight workers, and inline execution after disable. Green under `-race` on both gtk3 and gtk4.

File: v3/internal/assetserver/webview/mainthread_testsupport_linux.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+static void webview_test_quit_loop(void) {
+	webview_test_wait_running();
+	g_main_loop_quit(webview_test_loop);
+}
+
+// webview_test_on_loop_thread reports whether the caller is the thread that runs
+// the main loop.
+static int webview_test_on_loop_thread(void) {
+	return g_thread_self() == (GThread *)webview_test_loop_thread ? 1 : 0;
```
[the remark this answers is not in the document]
Comment (taliesin-ai) [reply v3/internal/assetserver/webview/mainthread_testsupport_linux.go]: Good catch — fixed in 160067c1e. All shared loop state (`webview_test_loop`, `webview_test_loop_thread`, and the readiness flag) is now written and read under a single `GMutex`. Readiness is published from inside the loop via an idle source that signals a `GCond`, so `testWaitLoopRunning` blocks on the cond instead of spinning on an unsynchronized flag, and `testOnLoopThread` reads the loop-thread handle under the lock. Still green under `-race` on both gtk3 and gtk4.

**Review by leaanthony — APPROVED** [review]
